src/sys/socket/mod.rs

```diff
@@ -2009,12 +2009,130 @@ pub fn socketpair<T: Into<Option<SockProtocol>>>(
     unsafe { Ok((OwnedFd::from_raw_fd(fds[0]), OwnedFd::from_raw_fd(fds[1]))) }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Backlog(i32);
+
+impl Backlog {
+    /// Sets the listen queue size to system `SOMAXCONN` value
+    pub const MAXCONN: Self = Self(libc::SOMAXCONN);
+    /// Sets the listen queue size to -1 for system supporting it
+    #[cfg(any(target_os = "linux", target_os = "freebsd"))]
+    pub const MAXALLOWABLE: Self = Self(-1);
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum BacklogTryFromError {
+    TooNegative,
+    TooPositive,
+}
+
+impl std::fmt::Display for BacklogTryFromError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            #[cfg(any(target_os = "linux", target_os = "freebsd"))]
+            Self::TooNegative => write!(f, "Passed a positive backlog less than -1"),
+            #[cfg(not(any(target_os = "linux", target_os = "freebsd")))]
+            Self::TooNegative => write!(f, "Passed a positive backlog less than 0"),
+            Self::TooPositive => write!(f, "Passed a positive backlog greater than `{:?}`", Backlog::MAXCONN)
+        }
+    }
+}
+
+impl std::error::Error for BacklogTryFromError {}
+
+impl<T: Into<Backlog>> From<Option<T>> for Backlog {
+    fn from(backlog: Option<T>) -> Self {
+        backlog.map_or(Self::MAXCONN, |b| b.into())
+    }
+}
+
+impl From<u16> for Backlog {
+    fn from(backlog: u16) -> Self {
+        Self(i32::from(backlog))
+    }
+}
+
+impl From<u8> for Backlog {
+    fn from(backlog: u8) -> Self {
+        Self(i32::from(backlog))
+    }
+}
```

Comment:
Just realized that we need to check the value even the value is in range of `i32`, so what about defining a `new(val: i32)` function:

```rs
/// The `backlog` argument of [`listen()`].
pub struct Backlog(i32);
impl Backlog {
    /// Sets the listen queue size to system `SOMAXCONN` value
    pub const MAXCONN: Backlog = Backlog(libc::SOMAXCONN);
    /// Sets the listen queue size to -1 for system supporting it
    #[cfg(any(target_os = "linux", target_os = "freebsd"))]
    pub const MAXALLOWABLE: Backlog = Backlog(-1);

    /// Create a `Backlog`, an `EINVAL` will be returned if `val` is invalid.
    pub fn new(val: i32) -> Result<Self> {
        cfg_if! {
            if #[cfg(any(target_os = "linux", target_os = "freebsd"))] {
                const MIN: i32 = -1;
            } else {
                const MIN: i32 = 0;
            }
        }

        if !(MIN..libc::SOMAXCONN).contains(&val) {
            return Err(Errno::EINVAL);
        }

        Ok(Self(val))
    }
}
```